fem/coefficient.cpp

```diff
@@ -170,6 +176,80 @@ void VectorGridFunctionCoefficient::Eval(
    GridFunc->GetVectorValues(T, ir, M);
 }
 
+GradientGridFunctionCoefficient::GradientGridFunctionCoefficient (
+   GridFunction *gf)
+   : VectorCoefficient((gf) ?
+                       gf -> FESpace() -> GetMesh() -> SpaceDimension() : 0)
+{
+   GridFunc = gf;
+}
+
+void GradientGridFunctionCoefficient::SetGridFunction(GridFunction *gf)
+{
+   GridFunc = gf; vdim = (gf) ?
+                         gf -> FESpace() -> GetMesh() -> SpaceDimension() : 0;
+}
+
+void GradientGridFunctionCoefficient::Eval(Vector &V, ElementTransformation &T,
+                                           const IntegrationPoint &ip)
+{
+   // Perhaps we should avoid altering T so cache its IntegrationPoint
+   const IntegrationPoint & ipT = T.GetIntPoint();
+   T.SetIntPoint(&ip);
+   GridFunc->GetGradient(T, V);
+   // Reset T's IntegrationPoint
+   T.SetIntPoint(&ipT);
+}
+
+void GradientGridFunctionCoefficient::Eval(
+   DenseMatrix &M, ElementTransformation &T, const IntegrationRule &ir)
+{
+   GridFunc->GetGradients(T.ElementNo, ir, M);
+}
```

Comment:
There maybe a potential issue here: if this method is called with the `ElementTransformation` returned by `Mesh::GetElementTransformation` then the call to the same method inside `GridFunction::GetGradients` becomes unnecessary. I think we can add a version of `GridFunction::GetGradients` that takes `ElementTransformation` instead of element index and use this new version here.

Reply:
That sounds reasonable.  I can make these changes.